[RISCV] add load/store misched/PostRA subtarget features

danielhb opened the pull request:
Some processors benefit more from store clustering than load clustering,
and vice-versa, depending on factors that are exclusive to each one
(e.g. macrofusions implemented).

Likewise, certain optimizations benefits more from misched clustering
than postRA clustering. Macrofusions are again an example: in a
processor with store pair macrofusions, like the veyron-v1, it is
observed that misched clustering increases the amount of macrofusions
more than postRA clustering. This of course isn't necessarily true for
other processors, but it shows that processors can benefit from a more
fine grained control of clustering mutations, and each one is able to do
it differently.

Add 4 new subtarget features that deprecates the existing
riscv-misched-load-store-clustering and riscv-postmisched-load-store-clustering
options:

- disable-misched-load-clustering and disable-misched-store-clustering:
disable load/store clustering during misched;

- disable-postmisched-load-clustering and disable-postmisched-store-clustering:
disable load/store clustering during PostRA.

Note that the new subtarget features disables specific stages of the default
clustering settings. The default per se (load and store clustering for both
misched and PostRA) is left untouched.

Disable all clustering but misched-store-clustering for the veyron-v1 processor
using the new features.

danielhb committed:
[RISCV] add load/store misched/PostRA cluster options

Some processors benefit more from store clustering than load clustering,
and vice-versa, depending on factors that are exclusive to each one
(e.g. macrofusions implemented).

Likewise, certain optimizations benefits more from misched clustering
than postRA clustering. Macrofusions are again an example: in a
processor with store pair macrofusions, like the veyron-v1, it is
observed that misched clustering increases the amount of macrofusions
more than postRA clustering.  This of course isn't necessarily true for
other processors, but it shows that processors can benefit from a more
fine grained control of clustering mutations, and each one is able to do
it differently.

Add 4 new clustering options that deprecates the existing
riscv-misched-load-store-clustering and
riscv-postmisched-load-store-clustering options:

- riscv-misched-load-clustering and riscv-misched-store-clustering:
  enable/disable load/store clustering during misched;

- riscv-postmisched-load-clustering and
  riscv-postmisched-store-clustering: enable/disable load/store
  clustering during PostRA.

To preserve the existing clustering behavior all 4 options are defaulted to
'true'.

mgudim commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -301,10 +321,13 @@ ScheduleDAGInstrs *
 RISCVTargetMachine::createMachineScheduler(MachineSchedContext *C) const {
   ScheduleDAGMILive *DAG = createSchedLive(C);
   if (EnableMISchedLoadStoreClustering) {
```
OK, got it. Thanks!

mgudim reviewed (approved):
LGTM, but let's wait to see what others say if it is OK to remove the old options.

mshockwave reviewed (commented):
I'm generally in favor of removing the existing two options (i.e. `riscv-misched-load-store-clustering` and `riscv-postmisched-load-store-clustering`). This is not a Clang flag and LLVM flags change all the time.

Non blocking, but an alternative solution to avoid having two flags could be making `-riscv-misched-load-store-clustering` (and its post-RA counterpart) to take a list of "load" or "store" instead to enable either or both clusterings.

mshockwave commented on llvm/test/CodeGen/RISCV/misched-load-clustering.ll:
could you add new tests checking the situation where only _either_ of the flags is enabled?

mshockwave reviewed (approved):
LGTM w/ minor comments

Please avoid force push unless necessary: https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

danielhb committed:
misched-load-clustering.ll: exercise new clustering flags

- add store-clustering=false to the existing LDCLUSTER test;
- add a new STCLUSTER test with load-clustering=false;
- add a new DEFAULTCLUSTER test with default clustering options.

wangpc-pp reviewed (approved):
LGTM! Thanks!

mgudim reviewed (approved):
LGTM, thanks!

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -94,14 +94,24 @@ static cl::opt<bool>
-static cl::opt<bool> EnableMISchedLoadStoreClustering(
-    "riscv-misched-load-store-clustering", cl::Hidden,
-    cl::desc("Enable load and store clustering in the machine scheduler"),
+static cl::opt<bool> EnableMISchedLoadClustering(
```
Did I miss something? We don't have subtarget features for these?

topperc replied:
That was my point. Should we add Subtarget features instead? The description of the PR suggests that whether this is profitable is CPU specific.

danielhb replied:
By Subtarget feature are we talking about a a feature, like FeatureUnalignedScalarMem, or a tune like TunePostRAScheduler (took the examples from sifive_p670)? Seems to me that this would be more like a tuning like PostRAScheduler.

Also, do we want to keep the default as is (enable all load/store clustering in both stages) and add features to disable it, or the other way around? Having everything disabled by default and adding subtarget features to enable what you want is cleaner, but  we'll change behavior for everyone that doesn't add the new features in their target defs. Perhaps this is a good thing (we'll force people to make a decision instead of relying on defaults) ....

danielhb committed:
Adding load/store clustering as subtarget features

Added 4 subtarget features that disables the default cluster settings:

- TuneDisableMISched(Load|Store)Clustering
- TuneDisablePostMISched(Load|Store)Clustering

The 4 command line options added previously were removed.

Use the new subtarget features in Ventana veyron-v1 to allow only
misched store clustering.

Test changes:

- add a store clustering test: misched-store-clustering.ll
- add subtarget features tests in misched-load-clustering.ll and
  misched-store-clustering.ll
- misched-mem-clustering.mir: use the subtarget features

mgudim commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1700,6 +1700,18 @@ def TuneNLogNVRGather
 def TunePostRAScheduler : SubtargetFeature<"use-postra-scheduler",
     "UsePostRAScheduler", "true", "Schedule again after register allocation">;
 
+def TuneDisableMISchedLoadClustering : SubtargetFeature<"disable-misched-load-clustering",
```
I think @topperc meant to mimic something like "bool RISCVSubtarget::useAA()" for example

topperc replied:
No I meant something we could set per-CPU like this.

danielhb committed:
Invert the enum logic like TuneNoDefaultUnroll does

Instead of using NoMISchedLoadClustering and setting it to 'true' using
the feature, rename it to NoMISchedLoadClustering and make the feature
set it to 'false'.

Rename the subtarget method disableMISchedLoadClustering() to
enableMISchedLoadClustering() to remove the negative check in
createMachineScheduler() when enabling the clustering.

Do the same with the other 3 subtarget features.

danielhb committed:
Remove 'enableXXXClustering' methods from RISCVSubtarget.h

We can get the existing enableXXClustering methods to be
autogenerated by prefixing the enums with 'Enable', e.g.
'EnableMISchedLoadClustering', and an autogenerated
'enableMISchedLoadClustering' method will be created.

wangpc-pp reviewed (approved):
LGTM.

topperc reviewed (approved):
LGTM

danielhb commented:
Ping

Do we require more reviews or is this good to go? Also, I don't have write access (this is my first PR for LLVM) so I'm afraid someone will have to do the honors merging this one.  I'm not sure if a rebase is needed due to the conflict being pointed out too ...

topperc commented:
> Ping
> 
> Do we require more reviews or is this good to go? Also, I don't have write access (this is my first PR for LLVM) so I'm afraid someone will have to do the honors merging this one. I'm not sure if a rebase is needed due to the conflict being pointed out too ...

You have enough reviews. The reviewers probably didn't think about you not having write permission.

mshockwave commented:
> > > Ping
> > > Do we require more reviews or is this good to go? Also, I don't have write access (this is my first PR for LLVM) so I'm afraid someone will have to do the honors merging this one. I'm not sure if a rebase is needed due to the conflict being pointed out too ...
> > 
> > 
> > You have enough reviews. The reviewers probably didn't think about you not having write permission.
> 
> Got it. I asked @mgudim for help.

I just merged it on your behalf
